### ISSUE TITLE
Bug in link click where elementClasses is `[null]`

### ISSUE THREAD

**Issue by alexanderdean** [snowplow/snowplow-javascript-tracker#417]
In js-2.5.2

```json
{\"schema\":\"iglu:com.snowplowanalytics.snowplow/unstruct_event/jsonschema/1-0-0\",\"data\":{\"schema\":\"iglu:com.snowplowanalytics.snowplow/link_click/jsonschema/1-0-1\",\"data\":{\"targetUrl\":\"http://www.efinancialcareers.sg/jobs-Information_Technology.s019\",\"elementId\":\"\",\"elementClasses\":[null],\"elementTarget\":\"\"}}}
```

Causes:

```
"errors":[{"level":"error","schema":{"loadingURI":"#","pointer":"/properties/elementClasses/items"},"instance":{"pointer":"/elementClasses/0"},"domain":"validation","keyword":"type","message":"instance type (null) does not match any allowed primitive type (allowed: [\"string\"])","found":"null","expected":["string"]}]}
```

**Comment by fblundun** [comment]
This bug must be intermittent - I haven't been able to reproduce it with either of the 2 links pointing to that page

**Comment by fblundun** [comment]
Even if the classList attribute for an element was somehow `null` or `undefined`, `lodash.map` just returns an empty array (`[]` not `[null]`)

**Comment by fblundun** [comment]
I considered the possibility that they might be directly using `trackLinkClick` and passing in `[null]`, but searching all scripts for "trackLinkClick" in Firebug shows this isn't the case.

**Comment by fblundun** [comment]
@alexanderdean how many bad link click events were there? And do you know what proportion of link clicks were bad?

**Comment by alexanderdean** [comment]
Not sure! Here is the useragent though:

    Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.0; Trident/5.0; yie9)\tInternet Explorer 9\tInternet Explorer\t9.0\tBrowser\tTRIDENT\ten-sg\t\t\t1\t\t\t\t\t\t\t1\t24\t1366\t714\tWindows Vista\tWindows\tMicrosoft Corporation

**Comment by fblundun** [comment]
According to the ua-parser enrichment, that corresponds to

```json
{
    "useragentFamily": "IE",
    "useragentMajor": "9",
    "useragentMinor": "0",
    "useragentPatch": null,
    "useragentVersion": "IE 9.0",
    "osFamily": "Windows Vista",
    "osMajor": null,
    "osMinor": null,
    "osPatch": null,
    "osPatchMinor": null,
    "osVersion": "Windows Vista",
    "deviceFamily": "Other"
}
```

I just tested using IE9 on a VM and again couldn't reproduce - the array was `[]`, not `[null]`.

**Comment by alexanderdean** [comment]
Very odd!

**Comment by fblundun** [comment]
Still don't know what caused this...

We could do a release just containing #418 and hopefully that will make the problem go away, since it would completely change the code used to extract the list of classes.

**Comment by alexanderdean** [comment]
Sounds like a plan to me - pushing this back to 2.6.1...

**Comment by fblundun** [comment]
Has this bug recurred since we did #418 in 2.5.3?

**Comment by fblundun** [comment]
Descheduling, since nobody has reproduced this and there is a good chance that it was fixed by 2.5.3.

**Comment by alexanderdean** [comment]
Let's close this then
